delira/training/predictor.py

```diff
@@ -0,0 +1,208 @@
+from .callbacks import AbstractCallback
+from batchgenerators.dataloading import MultiThreadedAugmenter
+import pickle
+from abc import abstractmethod
+import numpy as np
+from ..data_loading import BaseDataManager
+from tqdm import tqdm
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class Predictor(object):
+    """
+    Defines an API for Predictions from a Network
+
+    See Also
+    --------
+    :class:`PyTorchNetworkTrainer`
+
+    """
+
+    # static variable to prevent certain attributers from overwriting
+    __KEYS_TO_GUARD = []
+
+    def __init__(self, model, convert_batch_to_npy_fn=lambda *x: x,
+                 prepare_batch_fn=lambda *x: x, **kwargs):
+        """
+
+        Parameters
+        ----------
+        model :
+            model to predict from; must be callable
+
+        """
+        self._setup(model, convert_batch_to_npy_fn, prepare_batch_fn, **kwargs)
+
+    def _setup(self, network, convert_batch_to_npy_fn, prepare_batch_fn):
+        self.module = network
+        self._convert_batch_to_npy_fn = convert_batch_to_npy_fn
+        self._prepare_batch = prepare_batch_fn
+
+    def __call__(self, data):
+        return self.predict(data)
+
+    def predict(self, data, keys=["data"]):
+        data = self._prepare_batch(data)
+
+        if isinstance(data, dict):
+            data = [data[key] for key in keys]
+
+        return self._convert_batch_to_npy_fn(
+            self.module(
+                *data
+            )
+        )
+
+    def predict_data_mgr(self, datamgr, batchsize=None, verbose=False):
```

Comment:
in tf, a singleton dimension is added somewhere in here, leading to shape missmatches for inputs

Reply:
`classification_pytorch.ipynb` also fails with the following error message:

```RuntimeError: Expected 4-dimensional input for 4-dimensional weight [64, 1, 7, 7], but got 5-dimensional input of size [64, 1, 1, 224, 224] instead```

We again have a singleton dimension (and some additional shape missmatch) when calling `predictor.predict_data_mgr`
Real question is though: why are our tests not catching this?